src/main/drivers/compass_ist8310.c

```diff
@@ -0,0 +1,248 @@
+// #endif
+
+// #ifdef ENSURE_MAG_DATA_READY_IS_HIGH
+//     uint8_t status = GPIO_ReadInputDataBit(ist8310Config->gpioPort, ist8310Config->gpioPin);
+//     if (!status) {
+//         return;
+//     }
+// #endif
+
+//     registerExtiCallbackHandler(ist8310Config->exti_irqn, IST_MAG_DATA_READY_EXTI_Handler);
+
+//     EXTI_ClearITPendingBit(ist8310Config->exti_line);
+
+//     EXTI_InitTypeDef EXTIInit;
+//     EXTIInit.EXTI_Line = ist8310Config->exti_line;
+//     EXTIInit.EXTI_Mode = EXTI_Mode_Interrupt;
+//     EXTIInit.EXTI_Trigger = EXTI_Trigger_Falling;
+//     EXTIInit.EXTI_LineCmd = ENABLE;
+//     EXTI_Init(&EXTIInit);
+
+//     NVIC_InitTypeDef NVIC_InitStructure;
+
+//     NVIC_InitStructure.NVIC_IRQChannel = ist8310Config->exti_irqn;
+//     NVIC_InitStructure.NVIC_IRQChannelPreemptionPriority = NVIC_PRIORITY_BASE(NVIC_PRIO_MAG_DATA_READY);
+//     NVIC_InitStructure.NVIC_IRQChannelSubPriority = NVIC_PRIORITY_SUB(NVIC_PRIO_MAG_DATA_READY);
+//     NVIC_InitStructure.NVIC_IRQChannelCmd = ENABLE;
+//     NVIC_Init(&NVIC_InitStructure);
+// #endif
+// }
+
+bool ist8310Detect(mag_t* mag, const ist8310Config_t *ist8310ConfigToUse)
+{
+    bool ack = false;
+     uint8_t sig = 0;
+
+    ist8310Config = ist8310ConfigToUse;
+
+    ack = i2cRead(MAG_ADDRESS, MAG_WHOAMI, 1, &sig);
+    if (!ack || (sig != IST8310_ID))
+        return false;
+
+    mag->init = ist8310Init;
+    mag->read = ist8310Read;
+
+    return true;
+}
+
+void ist8310Init(void)
+{
+    int16_t magADC[3];
+    // int i;
+    // int32_t xyz_total[3] = { 0, 0, 0 }; // 32 bit totals so they won't overflow.
+    // bool bret = true;           // Error indicator
+
+    gpio_config_t gpio;
+
+    if (ist8310Config) {
+#ifdef STM32F303
+        if (ist8310Config->gpioAHBPeripherals) {
+            RCC_AHBPeriphClockCmd(ist8310Config->gpioAHBPeripherals, ENABLE);
+        }
+#endif
+#ifdef STM32F10X
+        if (ist8310Config->gpioAPB2Peripherals) {
+            RCC_APB2PeriphClockCmd(ist8310Config->gpioAPB2Peripherals, ENABLE);
+        }
+#endif
+        gpio.pin = ist8310Config->gpioPin;
+        gpio.speed = Speed_2MHz;
+        gpio.mode = Mode_IN_FLOATING;
+        gpioInit(ist8310Config->gpioPort, &gpio);
+    }
+
+    // 
+    i2cWrite(MAG_ADDRESS, IST8310_REG_CNTRL1, IST8310_SINGLE_MODE);
+    delay(5);
+    i2cWrite(MAG_ADDRESS, IST8310_AVERAGE, IST8310_AVG_16_TIME);
+    delay(5);
+    ist8310Read(magADC);
+    delay(5);
+    i2cWrite(MAG_ADDRESS, IST8310_REG_CNTRL1, IST8310_SINGLE_MODE);
+    
+    // 20160802 wait for ist8310 hardware pcb for data ready pin on stm32f303cc
+    // ist8310ConfigureDataReadyInterruptHandling();
+   
+}
+
+bool ist8310Read(int16_t *magData)
+{
+    uint8_t buf[6];
+    float LSB2FSV = 3; // 3mG - 14 bit
+    bool ack = i2cRead(MAG_ADDRESS, MAG_DATA_REGISTER, 6, buf);
+    if (!ack) {
+        return false;
+    }
+
+    // need to modify when confirming the pcb direction
+    magData[X] = -(int16_t)(buf[1] << 8 | buf[0]) * LSB2FSV;
```

Comment:
Does this mean that chip coordinate system is mirrored with respect to other compass chips?